File: module/node_bridge/nodey.js
import express from "express";
// import fetch from "node-fetch";

let app = express();

// Middleware to parse JSON
app.use(express.json());

let python_port = process.env.PYTHON_NODEY_PORT;

// app.post("/ask", async (req, res)=>{
//     let json = req.body();
//     json = json.asking_about;
//     try{
//         let response = await fetch("http://python:"+python_port+"/ask-python", {
//             method:"POST", 
//             body:JSON.stringify({asking_about:json})
//         });
//         let json = response.json();
//         res.send(json.data);
//     }
//     catch(e){
//         res.send(e);
//     }
// });

// app.post("/ask-node", (req, res)=>{
//     let json = req.body();
// });


export default class Nodey{
    constructor(){
        let port=process.env.NODE_NODEY_PORT;
        this.port = port;
        if (!port){
            throw new Error("Nodey reqires ports to run its bridge between python and node. No ports were given to run on.");
        }
        app.get("/", (req, res)=>{
            res.send("OK");
        });

        app.post("/", (req, res)=>{
            res.send("OK");
        });

        app.get('/health', (_req, res) => res.sendStatus(204));

        app.post("/ask", (req, res)=>{
            // console.log(req.body);
            let json = req.body;
            let token  = json.asking_about;
            // {"asking_about":"token"}
            let toSend;
            try{
            toSend = this.shareObject[token]();
            }
            catch(e){
                console.log(e);
                res.send(e);
            }
            let object_to_send = { data: Buffer.from(JSON.stringify(toSend)).toString('base64') };
            res.json(object_to_send);
        });
    }
    async ask(token){
        if (!token){
            throw new Error("Token is required to get data from python.");
        }
        const retries = 5;
        const delay = 2000; // milliseconds
        for (let i = 0; i < retries; i++) {
            try {
                // attempt fetch
                let response = await fetch("http://python:" + python_port + "/ask", {
                    method: "POST",
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({ asking_about: token })
                });
                let jsonResp = await response.json();
                let data = jsonResp.data;
                return Buffer.from(data, 'base64').toString();
            } catch (e) {
                if (i === retries - 1) {
                    throw new Error(`Nodey ask failed after ${retries} attempts: ${e}`);
                }
                await new Promise(res => setTimeout(res, delay));
            }
        }
    }
    set(shareObject){
        if(typeof shareObject!="object"){throw new Error("set() received bad arguments")}

        this.shareObject = shareObject;
        // {Custom_names_in_db:()=>{
        //  
        // }}
    }
    start(){
        app.listen(this.port, '0.0.0.0', () => {
            console.log(`Listening on port ${this.port}`);
        });
        
    }
}

File: module/python_bridge/nodey.py
from flask import Flask, request, jsonify
import os
import base64
import json
import time
import requests

app = Flask(__name__)

python_port = int(os.getenv("PYTHON_NODEY_PORT", 1931))
node_host = os.getenv("NODE_SERVICE", "http://node:8103")

class PyNodey:
    def __init__(self):
        self.port = python_port
        if not self.port:
            raise Exception("PyNodey requires PYTHON_NODEY_PORT to run.")

        self.share_object = {}

        @app.route('/health', methods=['GET'])
        def health():
            return '', 204

        @app.route("/", methods=["POST"])
        def base_response():
            return "OK", 200

        @app.route("/ask", methods=["POST"])
        def ask_handler():
            try:
                data = request.get_json()
                token = data.get("asking_about")
                if token not in self.share_object:
                    return jsonify({"error": "Invalid token"}), 400
                result = self.share_object[token]()
                encoded = base64.b64encode(json.dumps(result).encode()).decode()
                return jsonify({"data": encoded})
            except Exception as e:
                return jsonify({"error": str(e)}), 500

    def set(self, share_object):
        if not isinstance(share_object, dict):
            raise Exception("set() received bad arguments")
        self.share_object = share_object

    def ask(self, token):
        if not token:
            raise Exception("Token is required to get data from Node.")

        retries = 5
        delay = 2  # seconds
        last_exc = None
        for i in range(retries):
            try:
                response = requests.post(
                    f"{node_host}/ask",
                    json={"asking_about": token},
                    timeout=5
                )
                response.raise_for_status()
                response_json = response.json()
                encoded_data = response_json.get("data")
                decoded = base64.b64decode(encoded_data).decode()
                return json.loads(decoded)
            except Exception as e:
                last_exc = e
                if i < retries - 1:
                    time.sleep(delay)
        raise Exception(f"PyNodey ask failed after {retries} retries: {last_exc}")

    def start(self):
        print(f"PyNodey listening on port {self.port}")
        app.run(host="0.0.0.0", port=self.port)

# === Usage ===
if __name__ == "__main__":
    pynodey = PyNodey()

    pynodey.set({
        "Customers_in_db": lambda: "This is the data from Python",
        "Some_number": lambda: 1337,
        "Hello_obj": lambda: {"message": "Hello from Python!"}
    })

    pynodey.start()
